### PR TITLE
enable calling SaltSSH minions from the salt master

### LINKED ISSUE #40997: Allow Salt Master/Minion and Salt-ssh to work together
### Description of Issue/Question

It would be great to have salt-ssh and salt-minion/salt-master to be able to work together so they could cross call each other.

Would also need to make sure LocalClient can call across both so that https://github.com/saltstack/salt/issues/40943#issuecomment-297875083 can be fixed.

Daniel

### PR DESCRIPTION
**Pull request by gtmanfred**
### What does this PR do?
This adds the ability to let the salt-master spin off a new process which starts an SSHClient, and then returns the data back for the `salt` command to print it out.

### What issues does this PR fix or reference?
Closes #40997

### Tests written?

No

### Commits signed with GPG?

Yes

### CODE REVIEW

File: salt/master.py [hunk 1]
```diff
@@ -1909,6 +1910,14 @@ def publish(self, clear_load):
         payload = self._prep_pub(minions, jid, clear_load, extra, missing)
 
         # Send it!
+        if self.opts[u'enable_ssh_minions'] is True:
+            log.debug('Use SSHClient for rostered minions')
+            ssh = salt.client.ssh.client.SSHClient()
+            ssh_minions = ssh._prep_ssh(**clear_load).targets.keys()
+            if ssh_minions:
+                minions.extend(ssh_minions)
+                multiprocessing.Process(target=ssh.cmd, kwargs=clear_load).start()
```
Comment (thatch45) [inline salt/master.py]: @s0undt3ch This process will go zombie if I am not mistaken. Can the process manager be modified to auto clean up a process when it dies instead of restarting it?

**Comment by gtmanfred** [comment]
Once kitchen multinode is a thing, this would be another great use case for tests.

It looks like this is going to require a lot of mocks added to get it working with the test suite.

File: salt/master.py (hunk 1 above)
Comment (gtmanfred) [reply salt/master.py]: Just checked, and it does become a zombie.

**Comment by gtmanfred** [comment]
Any help on the last commit would be great, I can't figure out what is updating the jid in the `data` object, but if I just save it, and then reapply it, this makes the whole thing work with salt syndics as well.

File: salt/master.py [hunk 2]
```diff
@@ -1909,6 +1910,18 @@ def publish(self, clear_load):
...
+            # The isinstances makes sure that syndics work
+            log.debug('Use SSHClient for rostered minions')
+            ssh = salt.client.ssh.client.SSHClient()
+            ssh_minions = ssh._prep_ssh(**clear_load).targets.keys()
+            if ssh_minions:
+                minions.extend(ssh_minions)
+                def wrap_ssh(**kwargs):
+                    salt.utils.process.daemonize(False)
+                    ssh.cmd(**kwargs)
+                salt.utils.process.SignalHandlingMultiprocessingProcess(target=wrap_ssh, kwargs=clear_load).start()
```
Comment (gtmanfred) [inline salt/master.py]: This process is still becoming a zombie, even if it is daemonized, whatever runs the ssh connection is causing it to be zombied, possibly because of the generator that is created in `salt.client.ssh`, but if this is switched to threading the whole problem goes away.

@cachedout can you take a look.  Tom and I looked for a couple hours on Friday but couldn't figure out where to fix this.  We have the suspicion that the same thing also happens when running `salt-ssh` from the commandline, just that the zombie is cleaned up in a matter of milliseconds by the shell.

Thanks,
Daniel

**Comment by gtmanfred** [comment]
Go Go Jenkins!

**Comment by rallytime** [comment]
@gtmanfred Lint, please: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/16941/violations/file/salt/master.py/

**Comment by gtmanfred** [comment]
bah, fixed

**Comment by rallytime** [comment]
@gtmanfred Unfortunately, there's a merge conflict now. Can you rebase?

**Comment by gtmanfred** [comment]
done

**Review by thatch45 — APPROVED** [review]

**Comment by rallytime** [comment]
Ah, sorry @gtmanfred - this will need another rebase.

File: salt/client/ssh/__init__.py [hunk 3]
```diff
@@ -663,11 +663,34 @@ def run_iter(self, mine=False, jid=None):
             host = next(six.iterkeys(ret))
             self.cache_job(jid, host, ret[host], fun)
             if self.event:
+                _, data = next(six.iteritems(ret))
+                data['jid'] = jid  # make the jid in the payload the same as the jid in the tag
                 self.event.fire_event(
+<<<<<<< HEAD
```
Comment (cachedout) [inline salt/client/ssh/__init__.py]: You have unresolved merge issues in here.

**Review by cachedout — CHANGES_REQUESTED** [review]

**Comment by gtmanfred** [comment]
Yup, missed that one, just finished rebasing.

**Comment by gtmanfred** [comment]
Go Go Jenkins!

**Comment by rallytime** [comment]
Oh, unfortunately this has another merge conflict. Can you rebase one more time @gtmanfred?

**Comment by gtmanfred** [comment]
Done

**Comment by rallytime** [comment]
@gtmanfred There's some lint here: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/17375/violations/file/salt/config/__init__.py/

**Comment by gtmanfred** [comment]
Should be good now.

File: doc/ref/configuration/master.rst [hunk 4]
```diff
@@ -96,6 +96,23 @@ The user to run the Salt processes
 
 .. conf_master:: ret_port
 
+``enable_ssh_minions``
+----------------------
+
+
+Default: ``False``
+
+Tell the master to also use SaltSSH when running commands against minions.
```
Comment (cachedout) [inline doc/ref/configuration/master.rst]: This has always been `salt-ssh` and not `SaltSSH`. (Though I do like that. If we change it we should change it everywhere, however.)

File: doc/ref/configuration/master.rst [hunk 5]
```diff
@@ -96,6 +96,23 @@ The user to run the Salt processes
...
+
+.. code-block:: yaml
+
+    enable_ssh_minions: True
+
+.. note::
+
+    Cross minion type communication is still not possible.  The SaltMine and
```
Comment (cachedout) [inline doc/ref/configuration/master.rst]: `Cross-minion`. Probably remove the word `type` here.

Comment (cachedout) [inline doc/ref/configuration/master.rst]: `Salt mine`. Not `SaltMine`.

File: salt/client/ssh/__init__.py [hunk 6]
```diff
@@ -215,7 +215,7 @@ class SSH(object):
     def __init__(self, opts):
         self.__parsed_rosters = {SSH.ROSTER_UPDATE_FLAG: True}
         pull_sock = os.path.join(opts['sock_dir'], 'master_event_pull.ipc')
-        if os.path.isfile(pull_sock) and HAS_ZMQ:
+        if os.path.exists(pull_sock) and HAS_ZMQ:
```
Comment (cachedout) [inline salt/client/ssh/__init__.py]: Hmm, that's interesting. Why was this change necessary?

File: salt/master.py [hunk 7]
```diff
@@ -1957,6 +1959,15 @@ def publish(self, clear_load):
         payload = self._prep_pub(minions, jid, clear_load, extra, missing)
 
         # Send it!
+        if self.opts[u'enable_ssh_minions'] is True and isinstance(clear_load[u'tgt'], six.string_types):
```
Comment (cachedout) [inline salt/master.py]: This method (`publish`) has so much bloat, I think it would be best to split this out into a private method.

**Review by cachedout — CHANGES_REQUESTED** [review]

File: salt/client/ssh/__init__.py (hunk 6 above)
Comment (gtmanfred) [reply salt/client/ssh/__init__.py]: So, this never actually worked, os.path.isfile does not return true on sockets, so we were never firing these events on the event bus in the past.

That is why I just threw away the old event that would be thrown and changed it to be the same format as our regular minion events.

**Comment by rallytime** [comment]
@gtmanfred Lint :) https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/17474/violations/file/salt/master.py/

**Review by cachedout — APPROVED** [review]

**Comment by rallytime** [comment]
@gtmanfred One more rebase here and then we should be able to get this in. :)

**Comment by gtmanfred** [comment]
:+1: done
